common/client/multi_node.go

```diff
@@ -545,45 +556,202 @@ func (c *multiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OP
+// In the collection step, we gather as many results as possible while minimizing waiting time. This operation succeeds
+// on one of the following conditions:
+// * Received at least one success
+// * Received at least one result and `sendTxSoftTimeout` expired
+// * Received results from the sufficient number of nodes defined by sendTxQuorum.
+// The aggregation is based on the following conditions:
+// * If there is at least one success - returns success
+// * If there is at least one terminal error - returns terminal error
+// * If there is both success and terminal error - returns success and reports invariant violation
+// * Otherwise, returns any (effectively random) of the errors.
 func (c *multiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, RPC_CLIENT]) SendTransaction(ctx context.Context, tx TX) error {
-	main, nodeError := c.selectNode()
-	var all []SendOnlyNode[CHAIN_ID, RPC_CLIENT]
-	for _, n := range c.nodes {
-		all = append(all, n)
+	if len(c.nodes) == 0 {
+		return ErroringNodeError
 	}
-	all = append(all, c.sendonlys...)
-	for _, n := range all {
-		if n == main {
-			// main node is used at the end for the return value
-			continue
+
+	txResults := make(chan sendTxResult, len(c.nodes))
+	// Must wrap inside IfNotStopped to avoid waitgroup racing with Close
+	ok := c.IfNotStopped(func() {
+		c.wg.Add(len(c.sendonlys))
+		// fire-n-forget, as sendOnlyNodes can not be trusted with result reporting
+		for _, n := range c.sendonlys {
+			go c.broadcastTxAsync(ctx, n, tx, nil, &c.wg)
 		}
-		// Parallel send to all other nodes with ignored return value
-		// Async - we do not want to block the main thread with secondary nodes
-		// in case they are unreliable/slow.
-		// It is purely a "best effort" send.
-		// Resource is not unbounded because the default context has a timeout.
-		ok := c.IfNotStopped(func() {
-			// Must wrap inside IfNotStopped to avoid waitgroup racing with Close
-			c.wg.Add(1)
-			go func(n SendOnlyNode[CHAIN_ID, RPC_CLIENT]) {
-				defer c.wg.Done()
-
-				txErr := n.RPC().SendTransaction(ctx, tx)
-				c.lggr.Debugw("Sendonly node sent transaction", "name", n.String(), "tx", tx, "err", txErr)
-				sendOnlyError := c.sendOnlyErrorParser(txErr)
-				if sendOnlyError != Successful {
-					c.lggr.Warnw("RPC returned error", "name", n.String(), "tx", tx, "err", txErr)
-				}
-			}(n)
-		})
-		if !ok {
-			c.lggr.Debug("Cannot send transaction on sendonly node; MultiNode is stopped", "node", n.String())
+
+		// signal when all the primary nodes done broadcasting tx
+		inTxResults := make(chan sendTxResult, len(c.nodes))
+		var wg sync.WaitGroup
+		wg.Add(len(c.nodes))
+		c.wg.Add(1)
+		go func() {
+			wg.Wait()
+			c.wg.Done()
+			close(inTxResults)
+
+		}()
+
+		for _, n := range c.nodes {
+			go c.broadcastTxAsync(ctx, n, tx, inTxResults, &wg)
 		}
+
+		txResultsToReport := make(chan sendTxResult, len(c.nodes))
+		go fanOut(inTxResults, txResultsToReport, txResults)
```

Comment:
What if `broadcastTxAsync` is already over at this point and someone calls Close(), would we have a leak?

Reply:
`fanOut` is not leaking, because if `broadcastTxAsync` is already over - inTxResults is closed. Both `txResultsToReport`, `txResults` are sufficiently buffered, so we won't stuck even if no one is reading the results. 
Also `Close` can not complete before `fanOut ` is done as `reportSendTxAnomalies`, that is protected by a wait group, waits for `txResultsToReport` to be closed.

Reply:
I'm not suggesting that it will get stuck, I'm saying there is a rare case the multinode can close before the fanOut method returns. The fanOut method will still close eventually, perhaps almost instantly after the mutlinode close, but theoretically you can have a scenario where the parent routine exits before the child routine, which might get picked up by random tests, so perhaps it should be better if we do `c.wg.Add(2)` at the beginning.